File: 133-clone-graph/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) return node;

        unordered_map<Node*, Node*> m;

        Node *root = new Node(node->val, {});
        m[node] = root;

        queue<Node*> q;
        q.push(node);

        while(!q.empty()){
            Node *curr = q.front();
            q.pop();

            for(auto &adj:curr->neighbors){
                if(m.find(adj) == m.end()){
                    m[adj] = new Node(adj->val, {});
                    q.push(adj);
                }

                m[curr]->neighbors.push_back(m[adj]);
            }
        }

        return m[node];
    }
};

File: 133-clone-graph/clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
    if node == nil { return node }

    pair := make(map[*Node]*Node)

    var q []*Node
    q = append(q, node)
    for len(q) > 0 {
        curr := q[0]
        q = q[1:]

        newNode := &Node{}
        newNode.Val = curr.Val
        newNode.Neighbors = []*Node{}

        pair[curr] = newNode

        for _, n := range curr.Neighbors {
            if _, ok := pair[n]; ok { continue }
            q = append(q, n)
        }
    }

    for key, val := range pair {
        newNode := val
        var neighbors []*Node
        for _, v := range key.Neighbors {
            neighbors = append(neighbors, pair[v])
        }
        newNode.Neighbors = neighbors
    }

    return pair[node]
}